torch/csrc/dynamo/guards.cpp

```diff
@@ -674,6 +674,621 @@ static struct PyModuleDef _module = {
+       << "num_guards_executed=" << num_guards_executed << ")\n";
+    return ss.str();
+  }
+
+  // Whether the guard passed or failed.
+  bool result;
+
+  // This is a list of verbose_code_parts for the failed guard. When there are
+  // more than one verbose_code_parts, then recompilation reasoning infra on the
+  // Python side can iterate over this list and eval each string to pinpoint the
+  // exact code part that failed.
+  py::list verbose_code_parts;
+
+  // Total number of executed guards so far. This is helpful in debugging if
+  // shuffling is working.
+  int num_guards_executed;
+};
+
+/**
+ * Base class for the leaf guard in the GuardManager hierarchy.
+ */
+class LeafGuard {
+ public:
+  LeafGuard(py::object verbose_code_parts)
+      : _verbose_code_parts(verbose_code_parts) {}
+
+  // check function could be called from python. This is useful for debugging
+  // purpose.
+  bool check(py::handle value) {
+    return check_nopybind(value.ptr());
+  }
+
+  GuardDebugInfo check_verbose(py::handle value) {
+    return check_verbose_nopybind(value.ptr());
+  }
+
+  virtual GuardDebugInfo check_verbose_nopybind(
+      PyObject* value) { // borrowed ref
+    bool result = check_nopybind(value);
+    if (!result) {
+      return GuardDebugInfo(result, _verbose_code_parts, 0);
+    }
+    return GuardDebugInfo(true, 0);
+  }
+
+  py::list verbose_code_parts() {
+    return _verbose_code_parts;
+  }
+
+  // This is on the hot path and avoids any refcounting code from pybind. This
+  // is not exposed to Python and can only be called from C++.
+  virtual bool check_nopybind(PyObject* value) = 0;
+  virtual ~LeafGuard() = default;
+
+ private:
+  // This is set while constructing the leaf guard. This is used for identifying
+  // the cause of recompilation.
+  py::list _verbose_code_parts;
+};
+
+/**
+ * Represents a leaf guard that accepts the python guard check function. We
+ * would like to have most of the guards in C++ (to avoid a Python function
+ * call).  But, it will take some time to reach that goal. Also, there might be
+ * cases where its too tedious to write an equivalent C++ guard.
+ *
+ * LAMBDA_GUARD allows us to gradually move to C++. We can start from all
+ * guards of type PythonLambaGuard and incrementally move expensive guards to
+ * C++.
+ */
+class LAMBDA_GUARD : public LeafGuard {
+ public:
+  LAMBDA_GUARD(py::object guard_check_fn, py::object verbose_code_parts)
+      : LeafGuard(verbose_code_parts) {
+    if (py::isinstance<py::function>(guard_check_fn)) {
+      _guard_check_fn = py::cast<py::function>(guard_check_fn);
+    } else {
+      throw py::type_error("LAMBDA_GUARD expects (callable, str)");
+    }
+  }
+
+  // Runs the lambda function with the current f_locals value.
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    PyObject* x = PyObject_CallOneArg(_guard_check_fn.ptr(), value); // new ref
+    if (x == nullptr) {
+      // An exception is caught in the lambda function.
```

Comment:
If this is supposed to be unreachable should we assert rather than silently returning?

If we want to return you need to clear the Python error.

Reply:
Clearing the expection and returning False. I don't want to assert because otherwise then we might not DECREF some other pointer held higher up in the GuardManager tree.